docs/dev/how-to/estimate-gas.md

```diff
@@ -0,0 +1,399 @@
+```
+@tab Go
+```sh
+// In progress. Check back later.
+```
+@tab Java
+```sh
+// In progress. Check back later.
+```
+:::
+
+### Result
+
+6. The return value is a 256-bit unsigned integer in hexadecimal representing how much gas the transaction uses.
+
+```json
+{
+    "jsonrpc": "2.0",
+    "result": "0x25f64db",
+    "id": 2
+}
+```
+
+
+
+## L2 transactions
+
+Gas estimation for L2 to L2 transactions on zkSyncEra works in the same way as in Ethereum.
+
+Supply the same struct as for a L1 to L2 transaction.
+
+?? // can you add some code here please
+
+
+
+## L2 to L1 transactions
+
+Two transactions are required: 
+
+- An L2 transaction which sends a message of arbitrary length. 
+- When the block the message resides in is proved on L1, verification and confirmation takes place.
+
+For example, an L2 transaction uses a bridge to transfer funds. The tokens are removed from L2, and a message is sent to L1 saying the tokens were burned. On L1, the block containing the message is proved, and the bridge counterpart verifies that the token was indeed present in a certain block on L2 and supplies the funds.
+
+1. Import the zkSync Era library or contract containing the required functionality. 
+
+2. Get a `Contract` object that represents the [`Messenger`](../developer-guides/system-contracts.md#l1messenger) contract.
+
+3. Transform the request into a raw bytes array.
+
+4. Use the [`sendToL1`](https://github.com/matter-labs/v2-testnet-contracts/blob/b8449bf9c819098cc8bfee0549ff5094456be51d/l2/system-contracts/interfaces/IL1Messenger.sol#L5) function from the `IL1Messenger.sol` interface, passing the message as a raw bytes array.
+
+```solidity
+function sendToL1(bytes memory _message) external returns (bytes32);
+```
+
+:::warning What is a message?
+- A message functions like an event on Ethereum. The difference is that a message publishes data on L1. 
+- Its Solidity representation:
+    ```solidity
+    struct L2Message {
+            address sender;
+            bytes data;
+            uint256 txNumberInblock;
+    }
+    ```
+:::
+
+5. The [`proveL2MessageInclusion`](https://github.com/matter-labs/v2-testnet-contracts/blob/b8449bf9c819098cc8bfee0549ff5094456be51d/l1/contracts/zksync/facets/Mailbox.sol#L46) function returns a boolean parameter indicating whether the message was sent successfully to L1.
+
+```solidity
+function proveL2MessageInclusion(
+    uint256 _blockNumber,
+    uint256 _index,
+    L2Message memory _message,
+    bytes32[] calldata _proof
+
+) public view returns (bool) {
+    return _proveL2LogInclusion(_blockNumber, _index, _L2MessageToLog(_message), _proof);   
+}
+```
+
+### Example
+
+::: code-tabs
+@tab Solidity
+```solidity
+//SPDX-License-Identifier: Unlicense
+pragma solidity ^0.8.0;
+
+// Importing zkSync contract interface
+import "@matterlabs/zksync-contracts/l1/contracts/zksync/interfaces/IZkSync.sol";
+
+contract Example {
+
+    function sendMessageToL1() external returns(bytes32 messageHash) {
```

Comment:
This function should be present only on L2, it may be a bit confusing that the `Example` contract contains both L1 and L2 functions

Reply:
We say in the beginning of this section:

"Two transactions are required: 

- An L2 transaction which sends a message of arbitrary length. 
- When the block the message resides in is proved on L1, verification and confirmation takes place."

So there is one on L2, then one on L1 ... ?? We may need a quick chat :)

Reply:
OK, so the question is, when we say **two transactions** are required, are we also saying that these transactions cannot be in the same contract?

Reply:
we can reconvene on this one ..